src/headless/converse-muc.js

```diff
@@ -2361,7 +2361,17 @@ converse.plugins.add('converse-muc', {
                 const body = message.get('message');
                 if (!body) { return; }
                 if (u.isNewMessage(message) && this.isHidden()) {
-                    const settings = {'num_unread_general': this.get('num_unread_general') + 1};
+                    let first_unread = this.get('first_unread');
+
+                    if (this.get('num_unread_general') == 0) {
+                      if (first_unread) {
+                          const msg = this.messages.where({'msgid': first_unread});
+                          if (msg.length > 0) msg[0].set("first_unread", false);
+                      }
+                      message.set("first_unread", true);
+                      first_unread = message.get('msgid');
+                    }
+                    const settings = {'first_unread': first_unread, 'num_unread_general': this.get('num_unread_general') + 1};
```

Comment:
Looks to me like this code is mostly a copy/paste of what's in the `ChatBox` model. You can probably move the code in `ChatBox` into a new method and then call it from both places, instead of duplicating it in two places.

Reply:
There is slight difference in behavior. That is why you had to override the method in ChatRoom class. I will refactor into a re-usable method in ChatBox class provided the code does not get convoluted.

Reply:
Thanks. I implemented new method in ChatBox called setFirstUnreadMsgId.